File: sample-conference-videochat/sample-conference-videochat/Components/ChatScreen/ActionsMenuViewController/MenuViewController.m
//
//  MenuViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 2/6/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "MenuViewController.h"
#import "UITableView+Chat.h"
#import "MenuActionCell.h"
#import "UIView+Chat.h"

typedef void(^SelectedActionHandler)(void);

@interface MenuViewController () <UITabBarDelegate, UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIButton *cancelButton;
@property (nonatomic, strong) NSMutableArray<MenuAction *> *actions;
@end

@implementation MenuViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.tableView addShadowToTableViewWithShadowColor:[UIColor colorWithRed:0.78f green:0.81f blue:0.85f alpha:1.0f]];
    self.tableView.userInteractionEnabled = YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    [self setupViews];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    if (self.cancelAction) {
        self.cancelAction();
    }
}

- (void)setupViews {
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    switch (self.menuType) {
        case TypeMenuMediaInfo:
        case TypeMenuChatInfo:
            [self.tableView.rightAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.rightAnchor constant: -22.0f].active = YES;
            [self.tableView.widthAnchor constraintEqualToConstant: 154.0f].active = YES;
            [self.tableView.heightAnchor constraintEqualToConstant: self.tableView.contentSize.height].active = YES;
            break;
            
        case TypeMenuAppMenu:
            [self.tableView.leftAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leftAnchor constant: 22.0f].active = YES;
            [self.tableView.widthAnchor constraintEqualToConstant: 177.0f].active = YES;
            [self.tableView.heightAnchor constraintEqualToConstant: 280.0f].active = YES;
            break;

        default:
            break;
    }
    
    [self.tableView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant: 32.0f].active = YES;
    [self.tableView setRoundViewWithCornerRadius:6.0f];
    self.tableView.userInteractionEnabled = YES;
    [self.tableView reloadData];
}

- (IBAction)didTapcancelButton:(UIButton *)sender {
    [self dismissViewControllerAnimated:NO completion:^{
        if (self.cancelAction) {
            self.cancelAction();
        }
    }];
}

- (void)addAction:(MenuAction *)action {
    if (!self.actions) {
        self.actions = [NSMutableArray array];
    }
    [self.actions insertObject:action atIndex:0];
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.actions.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuActionCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MenuActionCell"];
    MenuAction *menuAction = self.actions[indexPath.row];
    cell.actionLabel.text = menuAction.title;

    if (self.menuType == TypeMenuAppMenu) {
        if (menuAction.action == ChatActionUserProfile) {
            cell.actionLabel.font = [UIFont systemFontOfSize:15.0f weight:UIFontWeightMedium];
        }
        if (indexPath.row == 1 || indexPath.row == self.actions.count - 1) {
            cell.separatorView.hidden = NO;
        }
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuAction *action = self.actions[indexPath.row];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self dismissViewControllerAnimated:NO completion:^{
            if (action.handler) {
                action.handler(action.action);
            }
            if (self.cancelAction) {
                self.cancelAction();
            }
        }];
    });
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuAction *menuAction = self.actions[indexPath.row];
    if (self.menuType == TypeMenuAppMenu) {
        if (menuAction.action == ChatActionLogout || menuAction.action == ChatActionUserProfile) {
            return 74.0f;
        }
    }
    return 44.0f;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/ChatScreen/ActionsMenuViewController/MenuViewController.swift
//
//  MenuViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 09.01.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit

enum TypeActionsMenuVC {
    case chatInfo
    case appMenu
    case mediaInfo
}

struct MenuConstant {
    static let chatInfoHeight: CGFloat = 182.0
    static let chatInfoWidth: CGFloat = 154.0
    static let appMenuHeight: CGFloat = 280.0
    static let appMenuWidth: CGFloat = 177.0
    static let heightForRow: CGFloat = 44.0
    static let heightForUserProfileRow: CGFloat = 74.0
}

class MenuViewController: UIViewController {
    @IBOutlet weak var containerView: UIView!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var cancelButton: UIButton!

    var cancelAction:(() -> Void)?
    private var actions:[MenuAction] = []
    var typeActionsMenuVC: TypeActionsMenuVC?

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        cancelAction?()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        setupViews()
    }
    
    private func setupViews() {
        guard let typePopVC = typeActionsMenuVC else {return}
        
        let heightConstant = tableView.contentSize.height + 6.0
        containerView.translatesAutoresizingMaskIntoConstraints = false
        containerView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32.0).isActive = true
        switch typePopVC {
        case .chatInfo:
            containerView.rightAnchor.constraint(equalTo: view.safeAreaLayoutGuide.rightAnchor, constant: -22.0).isActive = true
            containerView.widthAnchor.constraint(equalToConstant: MenuConstant.chatInfoWidth).isActive = true
            containerView.heightAnchor.constraint(equalToConstant: heightConstant).isActive = true
            containerView.addShadowToView(color: #colorLiteral(red: 0.7834715247, green: 0.8073117137, blue: 0.8447045684, alpha: 1))
        case .appMenu:
            containerView.leftAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leftAnchor, constant: 22.0).isActive = true
            containerView.widthAnchor.constraint(equalToConstant: MenuConstant.appMenuWidth).isActive = true
            containerView.heightAnchor.constraint(equalToConstant: MenuConstant.appMenuHeight).isActive = true
            containerView.addShadowToView(color: #colorLiteral(red: 0.7834715247, green: 0.8073117137, blue: 0.8447045684, alpha: 1))
        case .mediaInfo:
            containerView.rightAnchor.constraint(equalTo: view.safeAreaLayoutGuide.rightAnchor, constant: -22.0).isActive = true
            containerView.widthAnchor.constraint(equalToConstant: MenuConstant.chatInfoWidth).isActive = true
            containerView.heightAnchor.constraint(equalToConstant: heightConstant).isActive = true
        }
        
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 3.0).isActive = true
        tableView.rightAnchor.constraint(equalTo: containerView.rightAnchor).isActive = true
        tableView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: -3.0).isActive = true
        tableView.leftAnchor.constraint(equalTo: containerView.leftAnchor).isActive = true
        
        containerView.setRoundView(cornerRadius: 6.0)
        tableView.isUserInteractionEnabled = true
        tableView.reloadData()
    }
    
    func addAction(_ action: MenuAction) {
        actions.append(action)
    }
    
    @IBAction func tapCancelButton(_ sender: Any) {
        dismiss(animated: false) {
            self.cancelAction?()
        }
    }
}

extension MenuViewController: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return actions.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: MenuActionCellConstant.reuseIdentifier, for: indexPath) as? MenuActionCell else {
            return UITableViewCell()
        }
        
        let menuAction = actions[indexPath.row]
        cell.actionLabel.text = menuAction.title
        if let type = typeActionsMenuVC, type == .appMenu {
            if menuAction.action == .userProfile {
                cell.actionLabel.font = .systemFont(ofSize: 15.0, weight: .medium)
            }
            if indexPath.row == 1 || indexPath.row == actions.count - 1 {
                cell.separatorView.isHidden = false
            }
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let menuAction = actions[indexPath.row]
        DispatchQueue.main.async(execute: {
            self.dismiss(animated: false) {
                menuAction.successHandler?(menuAction.action)
                self.cancelAction?()
            }
        })
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let menuAction = actions[indexPath.row]
        if let type = typeActionsMenuVC, type == .appMenu {
            if menuAction.action == .logout || menuAction.action == .userProfile {
                return MenuConstant.heightForUserProfileRow
            }
        }
        return MenuConstant.heightForRow
    }
}
